torch/distributed/fsdp/sharded_grad_scaler.py

```diff
@@ -281,27 +281,41 @@ def unscale_(self, optimizer: torch.optim.Optimizer) -> None:
         # Synchronize the detected inf across the ranks
         optimizer_state = self._per_optimizer_states[id(optimizer)]
         future_handles = []
-
-        for v in optimizer_state["found_inf_per_device"].values():
-            if v.device.type == "cpu":
-                v_on_cuda = v.cuda()
+        found_inf_on_cudas = []
+
+        for found_inf in optimizer_state["found_inf_per_device"].values():
+            if found_inf.device.type == "cpu":
+                # found_inf is a scalar tensor on CPU
+                # it's small enough to live in pinned memory
+                # to speed up H2D transfer
+                found_inf_on_cuda = found_inf.pin_memory().cuda()
+                found_inf_on_cudas.append(found_inf_on_cuda)
                 future_handles.append(
                     dist.all_reduce(
-                        v_on_cuda, async_op=True, group=self.process_group
+                        found_inf_on_cuda, async_op=True, group=self.process_group
                     ).get_future()
                 )
-                v.copy_(v_on_cuda.cpu())
             else:
                 future_handles.append(
                     dist.all_reduce(
-                        v, async_op=True, group=self.process_group
+                        found_inf, async_op=True, group=self.process_group
                     ).get_future()
                 )
 
         # Make sure that the calls are done before moving out.
         if future_handles:
             torch.futures.wait_all(future_handles)
 
+        if found_inf_on_cudas:
+            for found_inf_on_cuda in optimizer_state["found_inf_per_device"].values():
+                if found_inf_on_cuda.device.type == "cpu":
+                    found_inf_on_cuda.copy_(found_inf_on_cudas.pop(0).cpu())
+
+        if found_inf_on_cudas:
```

Comment:
It might be cleaner to extract the values in `optimizer_state["found_inf_per_device"].values()` with `found_inf_on_cuda.device.type == "cpu"` into their own list that has the same length and can be indexed in lock-step with `found_inf_on_cudas`. This might make the logic clearer and avoid having this `if found_info_on_cudas:` check which should never return `True` and implicit reliance on `dict.values()` returning the same ordering.

Reply:
changed to lockstep forloop `zip(found_inf_on_cpus, found_inf_on_cudas)`